File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/sorting/Problem7453.java
package com.keunwon.algorithm.sorting;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Problem7453 {

    public static void main(String[] args) throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int[][] arr = new int[N][4];

        for (int i = 0; i < N; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());

            for (int j = 0; j < 4; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        int[] AB = new int[N * N];
        int[] CD = new int[N * N];
        int idx = 0;

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                AB[idx] = arr[i][0] + arr[j][1];
                CD[idx] = arr[i][2] + arr[j][3];
                idx++;
            }
        }

        Arrays.sort(AB);
        Arrays.sort(CD);

        long ans = 0;
        for (int i : AB) {
            int upper = upperBound(CD, -i);
            int lower = lowerBound(CD, -i);
            ans += (upper - lower);
        }

        System.out.println(ans);
    }

    static int lowerBound(int[] arr, int target) {
        int start = 0;
        int end = arr.length;

        while (start < end) {
            int mid = (start + end) / 2;

            if (arr[mid] < target) {
                start = mid + 1;
            } else {
                end = mid;
            }
        }
        return start;
    }

    static int upperBound(int[] arr, int target) {
        int start = 0;
        int end = arr.length;

        while (start < end) {
            int mid = (start + end) / 2;

            if (arr[mid] <= target) {
                start = mid + 1;
            } else {
                end = mid;
            }
        }
        return start;
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem7453.kt
package com.keunwon.algorithm.baekjoon

import java.util.*

/**
 * Title: 합이 0인 네 정수
 * Level: 골드-2
 **/
class Problem7453 {
    fun solution(arr: Array<IntArray>): Long {
        val cd = IntArray(arr.size * arr.size)
        var index = 0

        for (i in arr.indices) {
            for (j in arr.indices) {
                cd[index++] = arr[i][2] + arr[j][3]
            }
        }
        cd.sort()

        var answer = 0L
        for (i in arr.indices) {
            for (j in arr.indices) {
                val num = arr[i][0] + arr[j][1]
                val upperIndex = upperBound(-num, cd)
                val lowerIndex = lowerBound(-num, cd)

                answer += upperIndex - lowerIndex
            }
        }
        return answer
    }

    private fun lowerBound(target: Int, arr: IntArray): Int {
        var left = 0
        var right = arr.lastIndex

        while (left <= right) {
            val mid = (left + right) / 2

            if (target <= arr[mid]) right = mid - 1
            else left = mid + 1
        }
        return left
    }

    private fun upperBound(target: Int, arr: IntArray): Int {
        var left = 0
        var right = arr.lastIndex

        while (left <= right) {
            val mid = (left + right) / 2

            if (target < arr[mid]) right = mid - 1
            else left = mid + 1
        }
        return left
    }
}

fun main() = System.`in`.bufferedReader().use { br ->
    System.out.bufferedWriter().use { bw ->
        val n = br.readLine().toInt()
        val arr = Array(n) {
            val st = StringTokenizer(br.readLine())
            IntArray(4) { st.nextToken().toInt() }
        }

        Problem7453().solution(arr).also { bw.write("$it") }
        bw.flush()
    }
}
